### PR TITLE
Manual migrations API

### LINKED ISSUE #545: Migrations API

### PR DESCRIPTION
**Pull request by kristiandupont**
This is a new branch with cherrypicked commits from #642 and some fixes due to changes in master.
This resolves #545.

### CODE REVIEW

**Comment by kristiandupont** [comment]
@realm/dotnet this seems to work now, review please.

File: wrappers/src/shared_realm_cs.cpp [hunk 1]
```diff
@@ -52,92 +53,79 @@ class CSharpBindingContext: public BindingContext {
...
+    
+    bool read_only;
+    
+    bool in_memory;
+    
+    bool delete_if_migration_needed;
+
+    uint64_t schema_version;
+    
+     bool (*migration_callback)(SharedRealm* old_realm, SharedRealm* new_realm, SchemaForMarshaling, uint64_t schema_version, void* managed_migration_handle);
```
Comment (fealebenpae) [inline wrappers/src/shared_realm_cs.cpp]: whitespace

File: wrappers/src/shared_realm_cs.cpp [hunk 2]
```diff
@@ -52,92 +53,79 @@ class CSharpBindingContext: public BindingContext {
...
             config.schema_mode = SchemaMode::ResetFile;
         }
         
         config.schema = create_schema(objects, objects_length, properties);
-        config.schema_version = schemaVersion;
-
+        config.schema_version = configuration.schema_version;
+
+        if (configuration.managed_migration_handle) {
+            config.schema_mode = SchemaMode::Automatic;
```
Comment (fealebenpae) [inline wrappers/src/shared_realm_cs.cpp]: I believe it's `Automatic` by default.

**Comment by fealebenpae** [comment]
I think `SchemaHandle.cs` got included by mistake - it should actually remain deleted.

**Comment by fealebenpae** [comment]
Looks good to me, but that doesn't mean anything :)

**Comment by AndyDentFree** [comment]
No changelog entry?

File: Realm.Shared/Migration.cs [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+//
+////////////////////////////////////////////////////////////////////////////
+
+using System;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+
+namespace Realms
+{
+    public class Migration
```
Comment (AndyDentFree) [inline Realm.Shared/Migration.cs]: XML comment needed for API doc, of Migration class.

File: Realm.Shared/RealmConfiguration.cs [hunk 4]
```diff
@@ -89,6 +89,9 @@ public static string PathToRealm(string optionalPath = null)
         /// <value>0-based value initially set to zero so all user-set values will be greater.</value>
         public UInt64 SchemaVersion { get; set;} = 0;
 
+        public delegate void MigrationCallbackDelegate(Migration migration, UInt64 oldSchemaVersion);
```
Comment (AndyDentFree) [inline Realm.Shared/RealmConfiguration.cs]: XML comment needed for API doc here and on `MigrationCallback` property

File: Realm.Shared/handles/SchemaHandle.cs [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+
+namespace Realms
+{
+    internal class SchemaHandle : RealmHandle
+    {
+        private static class NativeMethods
+        {
+            [DllImport(InteropConfig.DLL_NAME, EntryPoint = "schema_create", CallingConvention = CallingConvention.Cdecl)]
+            public static extern unsafe IntPtr create([MarshalAs(UnmanagedType.LPArray), In] Native.SchemaObject[] objects, int objects_length,
+                                                              [MarshalAs(UnmanagedType.LPArray), In] Native.SchemaProperty[] properties,
```
Comment (AndyDentFree) [inline Realm.Shared/handles/SchemaHandle.cs]: indentation here and below

File: wrappers/src/shared_realm_cs.cpp [hunk 6]
```diff
@@ -52,92 +53,79 @@ class CSharpBindingContext: public BindingContext {
...
+    
+REALM_EXPORT SharedRealm* shared_realm_open(Configuration configuration, SchemaObject* objects, int objects_length, SchemaProperty* properties, uint8_t* encryption_key, NativeException::Marshallable& ex)
+{
+    return handle_errors(ex, [&]() {
+        Utf16StringAccessor pathStr(configuration.path, configuration.path_len);
 
         Realm::Config config;
         config.path = pathStr.to_string();
-        config.in_memory = durability != SharedGroup::durability_Full;
+        config.in_memory = configuration.in_memory;
```
Comment (AndyDentFree) [inline wrappers/src/shared_realm_cs.cpp]: We should also have a unit test to confirm we can create an in-memory Realm, now we are passing this setting through.

File: Realm.Shared/Realm.cs [hunk 7]
```diff
@@ -115,7 +115,30 @@ internal static Realm GetInstance(RealmConfiguration config, RealmSchema schema)
...
+            }
+
+            var srPtr = IntPtr.Zero;
+            try
+            {
+                srPtr = srHandle.Open(configuration, schema, config.EncryptionKey);
+            }
+            catch (ManagedExceptionDuringMigrationException)
+            {
+                throw new AggregateException("Exception occurred in a Realm migration callback. See inner exception for more details.", migration?.MigrationException);
```
Comment (AndyDentFree) [inline Realm.Shared/Realm.cs]: Strictly speaking, the _See inner exception for more details._ should be conditional on `migration != null`

File: Realm.Shared/handles/SchemaHandle.cs (hunk 5 above)
Comment (AndyDentFree) [inline Realm.Shared/handles/SchemaHandle.cs]: Scratch my indentation comment - as @fealebenpae said, suspect this entire file snuck back in during a merge.

File: wrappers/src/shared_realm_cs.cpp (hunk 6 above)
Comment (fealebenpae) [inline wrappers/src/shared_realm_cs.cpp]: We don't have a C# API that sets `in_memory`.

Comment (AndyDentFree) [inline wrappers/src/shared_realm_cs.cpp]: Noted - #191 is still P2

File: Realm.Shared/Realm.cs (hunk 7 above)
Comment (kristiandupont) [inline Realm.Shared/Realm.cs]: I think it's more a matter of the `?` operator being redundant. @fealebenpae is there any way for `migration` to be null when this exception is caught?

Comment (fealebenpae) [inline Realm.Shared/Realm.cs]: I don't think there is. We could safely drop it.

**Pull request merged**